plugins/bigquery/dbt/adapters/bigquery/impl.py

```diff
@@ -26,6 +29,90 @@
 
 import time
 import agate
+import re
+
+
+BQ_INTEGER_RANGE_NOT_SUPPORTED = f"""
+BigQuery integer range partitioning is only supported by the
+`partition_by` config, which accepts a dictionary.
+
+See: {dbt.links.BigQueryNewPartitionBy}
+"""
+
+
+@dataclass
+class PartitionConfig(JsonSchemaMixin):
+    field: str
+    data_type: str = 'date'
+    range: Optional[Dict[str, Any]] = None
+
+    def render(self, alias: Optional[str] = None):
+        column: str = self.field
+        if alias:
+            column = f'{alias}.{self.field}'
+
+        if self.data_type in ('timestamp', 'datetime'):
+            return f'date({column})'
+        else:
+            return column
+
+    @classmethod
+    def _parse(cls, raw_partition_by) -> Optional['PartitionConfig']:
+        if isinstance(raw_partition_by, dict):
+            try:
+                return cls.from_dict(raw_partition_by)
+            except ValidationError as exc:
+                msg = dbt.exceptions.validator_error_message(exc)
+                dbt.exceptions.raise_compiler_error(
+                    f'Could not parse partition config: {msg}'
+                )
+
+        elif isinstance(raw_partition_by, str):
+            raw_partition_by = raw_partition_by.strip()
+            if 'range_bucket' in raw_partition_by.lower():
+                dbt.exceptions.raise_compiler_error(
+                    BQ_INTEGER_RANGE_NOT_SUPPORTED
+                )
+
+            elif raw_partition_by.lower().startswith('date('):
+                matches = re.match(r'date\((.+)\)', raw_partition_by,
+                                   re.IGNORECASE)
+                if not matches:
+                    dbt.exceptions.raise_compiler_error(
+                        f"Specified partition_by '{raw_partition_by}' "
+                        "is not parseable")
+
+                partition_by = matches.group(1)
+                data_type = 'timestamp'
+
+            else:
+                partition_by = raw_partition_by
+                data_type = 'date'
+
+            inferred_partition_by = cls(
+                field=partition_by,
+                data_type=data_type
+            )
+
+            dbt.deprecations.warn(
+                'bq-partition-by-string',
+                raw_partition_by=raw_partition_by,
+                inferred_partition_by=inferred_partition_by
```

Comment:
Unsure if this is exactly the right way to do it, but can we render this out as a dict in the deprecation message?

![Screen Shot 2020-02-21 at 6 35 20 PM](https://user-images.githubusercontent.com/1541139/75080166-34482080-54d9-11ea-8505-22dc7243b12a.png)


```suggestion
                inferred_partition_by=inferred_partition_by.to_dict()
```